extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,298 @@
+  const start = Number(s);
+  let length = str.length;
+  user().assert(isFiniteNumber(start),
+      'Start index ' + start + 'in substr filter should be a number');
+  if (l) {
+    length = Number(l);
+    user().assert(isFiniteNumber(length),
+        'Length ' + length + ' in substr filter should be a number');
+  }
+
+  return str.substr(start, length);
+}
+
+/**
+ * @param {*} value
+ * @param {*} defaultValue
+ * @return {*}
+ */
+function defaultFilter(value, defaultValue) {
+  return value || user().assertString(defaultValue);
+}
+
+
+export class VariableService {
+  /**
+   * @param {!Window} window
+   */
+  constructor(window) {
+
+    /** @private {!Window} */
+    this.win_ = window;
+
+    /** @private {!Object<string, FilterDef>} */
+    this.filters_ = map();
+
+    this.register_('default', new FilterDef(defaultFilter, true));
+    this.register_('substr', new FilterDef(substrFilter));
+    this.register_('trim', new FilterDef(
+        value => user().assertString(value).trim()));
+    this.register_('json', new FilterDef(value => JSON.stringify(value)));
+    this.register_('toLowerCase', new FilterDef(value =>
+        user().assertString(value).toLowerCase()));
+    this.register_('toUpperCase', new FilterDef(value =>
+        user().assertString(value).toUpperCase()));
+    this.register_('not', new FilterDef(value => String(!value)));
+    this.register_('base64', new FilterDef(
+        value => btoa(user().assertString(value))));
+    this.register_('hash', new FilterDef(this.hashFilter_.bind(this)));
+    this.register_('if', new FilterDef(
+        (value, thenValue, elseValue) => value ? thenValue : elseValue, true));
+  }
+
+  /**
+   * @param {string} name
+   * @param {FilterDef} filter
+   */
+  register_(name, filter) {
+    dev().assert(!this.filters_[name], 'Filter "' + name
+        + '" already registered.');
+    this.filters_[name] = filter;
+  }
+
+  /**
+   * @param {string} filterStr
+   * @return {Object<string, FilterDef|Array<string>>}
+   */
+  parseFilter_(filterStr) {
+    if (!filterStr) {
+      return null;
+    }
+
+    // The parsing for filters breaks when `:` is used as something other than
+    // the argument separator. A full-fledged parser would be needed to fix
+    // this.
+    const tokens = filterStr.split(':');
+    const fnName = tokens.shift();
+    user().assert(fnName, 'Filter ' + name + ' is invalid.');
+    const filter = user().assert(this.filters_[fnName],
+        'Invalid filter name: ' + fnName);
+    return {filter, args: tokens};
+  }
+
+  /**
+   * @param {string} value
+   * @param {Array<string>} filters
+   * @return {Promise<string>}
+   */
+  applyFilters_(value, filters) {
+    let result = Promise.resolve(value);
+    for (let i = 0; i < filters.length; i++) {
+      const {filter, args} = this.parseFilter_(filters[i].trim());
+      if (filter) {
+        result = result.then(value => {
+          if (value != null || filter.allowNull) {
```

Comment:
We should just call the `#filter`, with the value and args, let it figure out if it should allow, then it'll call the filtering callback.